packages/flutter/lib/src/services/text_input.dart

```diff
@@ -710,7 +710,11 @@ class TextInputConnection {
   ///
   /// 2. [transform]: a matrix that maps the local paint coordinate system
   ///                 to the [PipelineOwner.rootNode].
+  ///
+  /// This method should only be called for web targets, and will result in an
+  /// exception on other platforms.
   void setEditableSizeAndTransform(Size editableBoxSize, Matrix4 transform) {
+    assert(kIsWeb);
```

Comment:
Do you think we can gracefully recover from the exception on Android side? When I talked to @Hixie  in summer, our philosophy was avoiding kIsWeb as much as possible.
I believe even if we add this as the next step the Android debugging will stop for setStyle. We will soon add other methods some already have design docs. Then in total this will make 3-4 'kIsWeb' checks just for text editing.

I believe the best approach is not to fail for optionalmethodcalls in the platforms.

Reply:
Yeah, i think the framework should just send all information down to the engine and the various engine implementations should decide what to do with that information. If some want to ignore it, that's fine.

Reply:
It's an exception to ignore a method call on a platform channel, and if we want to implement this we have to implement it for all embedders, and we have to update all of them to respond with a no-op.

Which means, in this case, we're making the render object get size and transformation info, encode it, send it, wait for a response that will ... do nothing. 

We could wrap this particular request in a `try {} on MissingPluginException {}`, but that seems worse to me than using kIsWeb.

Reply:
I think @dnfield's concern is valid. We shouldn't run unnecessary computations. But also we shouldn't put platform-specific branches in the code because that won't scale as we add more platforms. I think the right fix is for the engine to tell the framework what features it requires/supports. The simplest thing I can think of is a boolean flag in `dart:ui`, say, `requiresEditableGeometry`. Then we can do:

```dart
if (requiresEditableGeometry) {
  _textInputConnection.setEditableSizeAndTransform();
}
```

Each engine implementation will decide whether it needs editable geometry or not.

Reply:
Long-term, I'm not sure I'm a fan of the idea of the framework using hidden method channels. Their discoverability and documentation are extremely poor. I think the engine API surface should be entirely defined in `dart:ui` using plain Dart interfaces. Method channels should be left for plugins.

Reply:
@dnfield the text input channel is an `OptionalMethodChannel` and the framework already does what you suggested (see [here](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/platform_channel.dart#L427-L434)).

@yjbanov's suggestion could also work for this particular case, but I'm not sure that adding arbitrary booleans like this would work at scale. Maybe once this becomes a problem (and we notice a performance hit because of unnecessary platform-specific computations) we could come up with a general solution where `dart:ui` exposes all the channels and method names that it requires/supports.

Reply:
> I think the engine API surface should be entirely defined in dart:ui using plain Dart interfaces.

100% agree!

> Method channels should be left for plugins.

Why can't the same be done with plugins, I'm curious?

Reply:
Some method channels on the engine side for non-web-embedders are completey handled by Java, Objective C, or C++ code.  We could move the interface into dart:ui I suppose, but it makes dart:ui more complicated which we try to avoid.

Yes, I see now a try/catch wouldn't be enough. What we want to avoid is that someone doing a debugging session does not end up seeing exceptions thrown in the framework.  We also should really avoid using exceptions for control flow, as it's confusing (this isn't actually an exceptional circumstance).

Reply:
For now, I'm going to change things up a bit if I can to avoid having `OptionalMethodChannel` `throw` anything in this case.  

We can tackle enhancements to dart:ui as a separate issue.

Reply:
Do we need to fix OptionalMethodChannel than to not use exceptions as control flow?